snuba/datasets/tags_column_processor.py

```diff
@@ -77,38 +125,135 @@ def __string_col(self, col: str) -> str:
         else:
             return "toString({})".format(escape_identifier(col))
 
-    def __tag_expr(self, column_name: str, table_alias: str = "",) -> str:
+    def __tag_expr(self, parsed_col: ParsedNestedColumn, table_alias: str = "",) -> str:
         """
         Return an expression for the value of a single named tag.
 
         For tags/contexts, we expand the expression depending on whether the tag is
         "promoted" to a top level column, or whether we have to look in the tags map.
         """
-        col, tag = NESTED_COL_EXPR_RE.match(column_name).group(1, 2)
         # For promoted tags, return the column name.
+        assert parsed_col.parameter
+        tag_name = parsed_col.parameter
+        col = parsed_col.col_name
         if col in self.__promoted_columns:
-            actual_tag = self.__get_tag_column_map()[col].get(tag, tag)
+            actual_tag = self.__get_tag_column_map()[col].get(tag_name, tag_name)
             if actual_tag in self.__promoted_columns[col]:
                 return qualified_column(self.__string_col(actual_tag), table_alias)
 
         # For the rest, return an expression that looks it up in the nested tags.
-        return u"{col}.value[indexOf({col}.key, {tag})]".format(
-            **{"col": qualified_column(col, table_alias), "tag": escape_literal(tag)}
+        return "{col}.value[indexOf({col}.key, {tag})]".format(
+            **{
+                "col": qualified_column(col, table_alias),
+                "tag": escape_literal(tag_name),
+            }
+        )
+
+    def __extract_top_level_tag_conditions(self, condition: Expression) -> List[str]:
+        """
+        Finds the top level conditions that include tags_key in an expression.
+        For top level condition. In Snuba conditions are expressed in layers, the top level
+        ones are in AND, the nested ones are in OR and so on.
+        We can only apply the arrayFilter optimization to tag keys conditions that are not in
+        OR with other columns. To simplify the problem, we only consider those conditions that
+        are included in the first level of the query:
+        [['tagskey' '=' 'a'],['col' '=' 'b'],['col2' '=' 'c']]  works
+        [[['tagskey' '=' 'a'], ['col2' '=' 'b']], ['tagskey' '=' 'c']] does not
+        """
+
+        if is_binary_condition(condition, ConditionFunctions.EQ):
+            # This is to make mypy happy.
+            assert isinstance(condition, FunctionCall)
```

Comment:
I am on the verge of adding some wrappers around specific types of FunctionCall to make them easier to handle. I am not sure yet whether to add conditions as a first class element in the AST or to write separate lightweight wrappers to make manipulation easier. I think I will look how this work for a while and I will make the change when we will have a few more query processors based on the new AST.

Reply:
I don't really understand this comment, mostly because I'm having trouble projecting what these changes would actually look like concretely.

Broadly speaking, I think we should avoid adding to the AST object model as much as possible. (My knee jerk reaction is that I don't think a condition is appropriate at that level, but that's hard to really say without seeing any implementation or more at least understanding what you mean here in more detail.) I think it's fine to have classes that generate (or are interpretable as, maybe via subclassing) AST objects, but I think we should treat these as extensions and not as part of the AST itself.

Reply:
I agree, I am not planning to add anything to the AST. When I was talking about wrappers I was talking about classes/methods to help inspect the AST (more than generate the ast). But I am not adding anything at this point anyway.